fix: OnClientDisconnected client identifier is incorrect when pending client connection is denied [MTT-6376]

NoelStephensUnity opened the pull request:
This resolves the issue where the server-host was not receiving the correct client identifier when OnClientDisconnected is invoked if a pending client is denied approval.

[MTT-6376](https://jira.unity3d.com/browse/MTT-6376)
Resolves issue #2566 


## Changelog

- Fixed: issue where the `OnClientDisconnected` client identifier was incorrect when trigger after a pending client connection was denied.


## Testing and Documentation

- Includes integration test `ClientApprovalDenied.ClientApprovalDeniedNotificationTest`.
- No documentation changes or additions were necessary.

NoelStephensUnity committed:
fix

When connection approval is enabled and a client is not approved, the OnClientDisconnected callback handler was not being invoked.
(It was when a client terminates the connection but not when the server does)

NoelStephensUnity committed:
test

The test to validate the fix for the server-host not receiving the correct client identifier when denied approval for its pending connection.
Added additional functionality to NetcodeIntegrationTest that allows you to ignore waiting for a newly created client (i.e. not the default defined NumberOfClients) to connect.

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
style

grammar update

ShadauxCat commented on com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTest.cs:
```diff
@@ -431,6 +431,18 @@ protected virtual void OnNewClientStartedAndConnected(NetworkManager networkMana
+        /// <remarks>
+        /// Use this for testing connection and disconnection scenarios
+        /// </remarks>
+        protected virtual bool WaitForNewClientToConnect(NetworkManager networkManager)
```
This feels like it should be a property so it's more obvious the intent. When I first looked at this, I thought "this is just returning true, it's not waiting for a new client to connect at all".

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkConnectionManager.cs:
```diff
@@ -844,6 +844,16 @@ internal void OnClientDisconnectFromServer(ulong clientId)
+
+                try
+                {
+                    OnClientDisconnectCallback?.Invoke(clientId);
```
Just want to understand here...

Based on the description, OnClientDisconnectCallback was getting invoked with the wrong client ID previously. I can see how this would invoke it with the correct ID... but I don't see any changes to stop the previous invocation from calling with the wrong ID... so won't this result in it getting called twice, once with the correct ID and once with the wrong one?

NoelStephensUnity replied:
The other one is not called if the host disconnects the client from its side.
The recent changes with the NetworkConnectionManager actually wouldn't invoke OnClientDisconnected under this scenario at all.

NoelStephensUnity committed:
style

renaming NetcodeIntegrationTest.WaitForNewClientToConnect to NetcodeIntegrationTest.ShouldWaitForNewClientToConnect

NoelStephensUnity committed:
update

missed a rename of the virtual method invocation.

NoelStephensUnity committed:
test

Adding additional validation to assure we don't get multiple disconnect notifications on the server-host side when a client is disconnected.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
style

netcode-ci-service committed:
Merge develop into fix/client-disconnect-callback-after-denied
